c10/core/impl/alloc_cpu.cpp

```diff
@@ -70,6 +70,31 @@ void* alloc_cpu(size_t nbytes) {
       "DefaultCPUAllocator: not enough memory: you tried to allocate ",
       nbytes,
       " bytes.");
+#elif defined(__linux__)
+  int err = 0;
+  if (nbytes >= gAlloc_threshold_thp) {
+    // inorder to enable thp, buffers need to be page aligned
+    err = posix_memalign(&data, gAlignment_thp, nbytes);
+  } else {
+    err = posix_memalign(&data, gAlignment, nbytes);
+  }
+  CAFFE_ENFORCE(
+      err == 0,
+      "DefaultCPUAllocator: can't allocate memory: you tried to allocate ",
+      nbytes,
+      " bytes. Error code ",
+      err,
+      " (",
+      strerror(err),
+      ")");
+  // enable thp (transparent huge pages) for larger buffers
+  if (nbytes >= gAlloc_threshold_thp) {
+    int ret = madvise(data, nbytes, MADV_HUGEPAGE);
+    if (ret != 0) {
+      TORCH_WARN_ONCE(
+          "thp madvise for HUGEPAGE failed with %s\n", strerror(errno));
+    }
```

Comment:
Can we make it as a user opt-in option to enable THP on large memory allocation? THP reduces TLB miss rate but might have side effects like larger memory consumption, more fragmentations, kernel defragmentation overhead, perf fluctuation of kernel clear_page. The benefit would be workload specific and with tradeoffs.

Reply:
one option to disable the thp system wide is by setting
`echo never > /sys/kernel/mm/transparent_hugepage/enabled`
If there is any runtime prop based feature enable/disable in pytorch please let me know, i will try to use that for selectively disabling it.

Reply:
I was referring to what is the default behavior without additional user settings. For the system-wide setting, `madvise` is the default. For PyTorch, I suggested we make this an opt-in, not enabled by default. One option is to have a frontend API for users to enable explicitly.

Reply:
Hi @jgong5 , I agree with having configurability a better option, but I'm thinking how well user be made aware of it.
if you could point me to any other feature currently being controlled via frontend API, I would like to understand how they get used today. thank you!

Reply:
Currently, there are some control flags under `torch.backends.xyz` as well as other places for example, e.g.,
`torch.backends.mkldnn.*`: https://github.com/pytorch/pytorch/blob/11f51e798f480680a7634fbee2c3fb3cea09f970/torch/backends/mkldnn/__init__.py#L75, and `torch.backends.cudnn.*` etc.

My two cents on the places to add the flag. One option is to add it under `torch.cpu`, e.g., `torch.cpu.mm.transparent_hugepage.enabled`.

Reply:
thank you! will take a look.